prov/verbs/src/verbs_cq.c

```diff
@@ -230,47 +230,202 @@ static void fi_ibv_cq_read_data_entry(struct ibv_wc *wc, int i, void *buf)
+	struct fi_ibv_wce *wce;
+	struct util_buf_pool *wre_pool;
+	struct ibv_wc wc;
+	ssize_t ret;
+
+	ret = ibv_poll_cq(cq->cq, 1, &wc);
+	if (ret <= 0)
+		return ret;
+
+	if ((wc.opcode == IBV_WC_RECV) ||
+	    (wc.opcode == IBV_WC_RECV_RDMA_WITH_IMM) ||
+	    ((wc.wr_id & cq->wr_id_mask) != cq->send_signal_wr_id)) {
+		wre = (struct fi_ibv_wre *)(uintptr_t)wc.wr_id;
+		if (wre->ep) {
+			wre_pool = wre->ep->wre_pool;
+			if ((wre->ep != ep) &&
+			    (wc.status != IBV_WC_WR_FLUSH_ERR)) {
+				ret = fi_ibv_wc_2_wce(cq, &wc, &wce);
+				if (ret) {
+					wre->ep = NULL;
+					ret = -FI_EAGAIN;
+					goto fn;
+				}
+				slist_insert_tail(&wce->entry, &cq->wcq);
+			}
+			wre->ep = NULL;
+		} else {
+			/* WRE belongs to SRQ's wre pool and should be
+			 * handled or rejected if status ==  */
+			assert(wre->srq);
+			wre_pool = wre->srq->wre_pool;
+			wre->srq = NULL;
+			if (wc.status != IBV_WC_WR_FLUSH_ERR) {
+				ret = fi_ibv_wc_2_wce(cq, &wc, &wce);
+				if (ret) {
+					ret = -FI_EAGAIN;
+					goto fn;
+				}
+				slist_insert_tail(&wce->entry, &cq->wcq);
+			}
+		}
+fn:
+		dlist_remove(&wre->entry);
+		util_buf_release(wre_pool, wre);
+		return ret;
+	} else {
+		ret = fi_ibv_handle_internal_signal_wc(cq, &wc);
+		return ret ? ret : 1;
+	}
+}
+
+void fi_ibv_empty_wre_list(struct util_buf_pool *wre_pool,
+			   struct dlist_entry *wre_list,
+			   enum fi_ibv_wre_type wre_type)
+{
+	struct fi_ibv_wre *wre;
+	struct dlist_entry *tmp;
+
+	dlist_foreach_container_safe(wre_list, struct fi_ibv_wre,
+				     wre, entry, tmp) {
+		if (wre->wr.type == wre_type) {
+			dlist_remove(&wre->entry);
+			wre->ep = NULL;
+			wre->srq = NULL;
+			util_buf_release(wre_pool, wre);
+		}
+	}
+}
+
+void fi_ibv_cleanup_cq(struct fi_ibv_msg_ep *ep)
+{
+	int ret;
+
+	fastlock_acquire(&ep->rcq->lock);
+	do {
+		ret = fi_ibv_poll_outstanding_cq(ep, ep->rcq);
+	} while (ret > 0);
+
+	/* Handle WRs for which there were no appropriate WCs */
+	fi_ibv_empty_wre_list(ep->wre_pool, &ep->wre_list, IBV_RECV_WR);
+	fastlock_release(&ep->rcq->lock);
+
+	fastlock_acquire(&ep->scq->lock);
+	do {
+		ret = fi_ibv_poll_outstanding_cq(ep, ep->scq);
+	} while (ret > 0);
+	/* Handle WRs for which there were no appropriate WCs */
+
+	fi_ibv_empty_wre_list(ep->wre_pool, &ep->wre_list, IBV_SEND_WR);
+	fastlock_release(&ep->scq->lock);
+
+	util_buf_pool_destroy(ep->wre_pool);
+}
+
+/* Must call with cq->lock held */
+ssize_t fi_ibv_poll_cq(struct fi_ibv_cq *cq, struct ibv_wc *wc)
+{
+	struct fi_ibv_wre *wre;
+	struct util_buf_pool *wre_pool;
+	ssize_t ret;
+
+	ret = ibv_poll_cq(cq->cq, 1, wc);
+	if (ret <= 0)
+		return ret;
+
+	/* TODO Handle the case when app posts a send with same wr_id */
+	if ((wc->opcode == IBV_WC_RECV) ||
+	    (wc->opcode == IBV_WC_RECV_RDMA_WITH_IMM) ||
+	    ((wc->wr_id & cq->wr_id_mask) != cq->send_signal_wr_id)) {
+		wre = (struct fi_ibv_wre *)(uintptr_t)wc->wr_id;
+		if (!wre->ep && !wre->srq) {
+			/* it means that WRE has already been released */
```

Comment:
This case can never happen, right? Once a work completion is read, we release wre back to buf pool.

Reply:
Yes, you're right. Thanks!

Reply:
I'll put `assert` here.
It will be nice trigger for further changes if someone misses cq::lock keeping or smth else